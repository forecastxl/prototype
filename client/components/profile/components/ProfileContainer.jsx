import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import * as actions from '../actions'
import * as types from '../actionTypes'
import Profile from './Profile'

export class ProfileContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      profile: {
        firstName: '',
        lastName: '',
        email: '',
        currentPassword: '',
        newPassword: '',
        newPasswordConfirmation: ''
      }
    }
    this.onChange = this.onChange.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
  }

  componentDidMount() {
    this.props.fetchProfile()
  }

  componentWillReceiveProps({ profile }) {
    if (profile) {
      this.setState({ profile: {
        ...this.state.profile,
        firstName: profile.firstName,
        lastName: profile.lastName,
        email: profile.email
      } })
    }
  }

  onChange(event) {
    this.setState({
      profile: {
        ...this.state.profile,
        [event.target.name]: event.target.value
      }
    })
  }

  onSubmit(event) {
    event.preventDefault()

    const user = {
      ...this.state.profile,
      id: this.props.profile.id
    }

    this.props.updateProfile(user)
  }

  render() {
    return (
      <Profile
        profile={this.state.profile}
        onChange={this.onChange}
        onSubmit={this.onSubmit}
        errors={this.props.errors}
      />
    )
  }
}

ProfileContainer.propTypes = {
  errors: PropTypes.object.isRequired,
  fetchProfile: PropTypes.func.isRequired,
  updateProfile: PropTypes.func.isRequired,
  profile: PropTypes.shape({
    id: PropTypes.number.isRequired,
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired
  }).isRequired
}

const mapActions = {
  fetchProfile: actions.fetchProfile,
  updateProfile: actions.updateProfile
}

const mapStateToProps = state => ({
  errors: state.api[types.UPDATE_PROFILE].errors,
  profile: state.profile
})

export default connect(mapStateToProps, mapActions)(ProfileContainer)
