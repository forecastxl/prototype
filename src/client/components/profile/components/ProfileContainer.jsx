import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import { getProfile } from '../selectors'
import * as actions from '../actions'
import Profile from './Profile'

export class ProfileContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      profile: {
        firstName: '',
        lastName: '',
        email: '',
        currentPassword: '',
        newPassword: '',
        newPasswordConfirmation: ''
      }
    }
    this.onChange = this.onChange.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
  }

  componentDidMount() {
    this.props.fetchProfile()
  }

  componentWillReceiveProps({ profile }) {
    if (profile) {
      this.setState({
        profile: {
          ...this.state.profile,
          firstName: profile.firstName,
          lastName: profile.lastName,
          email: profile.email
        }
      })
    }
  }

  onChange(event) {
    this.setState({
      profile: {
        ...this.state.profile,
        [event.target.name]: event.target.value
      }
    })
  }

  onSubmit(event) {
    event.preventDefault()

    const user = {
      ...this.state.profile,
      id: this.props.profile.id
    }

    this.props.updateProfile(user)
  }

  render() {
    return (
      <Profile
        profile={this.state.profile}
        onChange={this.onChange}
        onSubmit={this.onSubmit}
        errors={{}}
      />
    )
  }
}

ProfileContainer.propTypes = {
  fetchProfile: PropTypes.func.isRequired,
  updateProfile: PropTypes.func.isRequired,
  profile: PropTypes.shape({
    id: PropTypes.number.isRequired,
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired
  }).isRequired
}

const mapActions = {
  fetchProfile: actions.fetchProfile,
  updateProfile: actions.updateProfile
}

const mapStateToProps = state => ({
  profile: getProfile(state)
})

export default connect(mapStateToProps, mapActions)(ProfileContainer)
