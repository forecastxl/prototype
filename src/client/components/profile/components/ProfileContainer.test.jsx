import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import { ProfileContainer } from './ProfileContainer'

jest.mock('material-ui/TextField')
jest.mock('../../alert')
jest.mock('../../spacer')
jest.mock('../../dialog')
jest.mock('../../container')

describe('<ProfileContainer />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <ProfileContainer
        fetchProfile={() => {}}
        updateProfile={() => {}}
        errors={{}}
        profile={{
          id: 1,
          firstName: 'Geert',
          lastName: 'Geerts',
          email: 'geert@email.com'
        }}
      />
    )
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('updates profile state after onChange call', () => {
    const wrapper = shallow(
      <ProfileContainer
        fetchProfile={() => {}}
        updateProfile={() => {}}
        errors={{}}
        profile={{
          id: 1,
          firstName: 'Geert',
          lastName: 'Geerts',
          email: 'geert@email.com'
        }}
      />
    )
    const event = {
      target: {
        name: 'firstName',
        value: 'updated'
      }
    }
    const expected = {
      ...wrapper.state('profile'),
      [event.target.name]: event.target.value
    }

    wrapper.instance().onChange(event)
    expect(wrapper.state('profile')).toEqual(expected)
  })

  it('processes submits', () => {
    const wrapper = mount(
      <ProfileContainer
        fetchProfile={() => {}}
        updateProfile={jest.fn()}
        errors={{}}
        profile={{
          id: 1,
          firstName: 'Geert',
          lastName: 'Geerts',
          email: 'geert@email.com'
        }}
      />
    )
    const event = {
      preventDefault: jest.fn()
    }
    const profileWithId = {
      ...wrapper.state('profile'),
      id: 1
    }

    wrapper.instance().onSubmit(event)
    expect(event.preventDefault).toHaveBeenCalled()
    expect(wrapper.prop('updateProfile')).toHaveBeenCalledWith(profileWithId)
  })

  it('handles incoming props', () => {
    const oldProfile = {
      currentPassword: '',
      newPassword: '',
      newPasswordConfirmation: '',
      firstName: 'Geert',
      lastName: 'Geerts',
      email: 'geert@email.com'
    }
    const newProfile = {
      firstName: 'John',
      lastName: 'Johnson',
      email: 'john@john.com'
    }
    const wrapper = mount(
      <ProfileContainer
        fetchProfile={() => {}}
        updateProfile={() => {}}
        errors={{}}
        profile={{ ...oldProfile, id: 1 }}
      />
    )
    wrapper.setProps({
      profile: { ...newProfile, id: 2 }
    })

    expect(wrapper.state('profile')).toEqual({ ...oldProfile, ...newProfile })
  })
})
