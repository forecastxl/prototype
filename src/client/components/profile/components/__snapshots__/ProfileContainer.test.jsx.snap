// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileContainer /> renders correctly 1`] = `
<Profile
  errors={Object {}}
  onChange={[Function]}
  onSubmit={[Function]}
  profile={
    Object {
      "currentPassword": "",
      "email": "",
      "firstName": "",
      "lastName": "",
      "newPassword": "",
      "newPasswordConfirmation": "",
    }
  }
/>
`;
